rfcs/2025-shadow-dom-support.md

```diff
@@ -0,0 +1,128 @@
+
+One proposal to avoid this traversal is to create focusable sentinels in order to trap focus.
+
+This will have the side benefit of theoretically working with closed root shadow doms as well, such as the aforementioned native video players inside a Dialog.
+
+3. In-team Education
+
+Writing code that can handle the prescence of Shadow DOM can be tricky. As a result, there will be a learning curve for the team.
+
+Some work has already been done to write utilities to hide away some of this complexity. We will also want to write lint rules to help us automatically catch as many common situations as possible. Some examples of these are:
+
+```jsx
+document.activeElement
+<->
+getActiveElement();
+
+e.target
+<->
+getTargetElement(e);
+
+let walker = document.createTreeWalker(...)
+<->
+let walker = createShadowTreeWalker(...)
+
+e.currentTarget.contains(e.target)
+<->
+nodeContains(e.currentTarget, getEventTarget(e.nativeEvent))
+```
+
+4. Communication
+
+We have our current Shadow DOM support gated behind a feature flag to both signal that it is experimental as well as to protect users who are not using any Shadow DOM from possible performance hits.
+
+We will want to expand this to also shield those users from code size increases. This could be accomplished with a global registry for the utility functions we provide that would replace the feature flag system we have for it right now.
+
+We also need to communicate, through documentation, what the specifics are of our Shadow DOM support since we likely will not be able to support everything, and certainly not much for a while.
+
+For example, some of our support may require open root Shadow DOMs. Or someone using our FocusScope's focus manager may need to know about different traversal methods to avoid the same issues we are trying to avoid.
+
+## Drawbacks
+
+One concern is on-going support. Our current use cases do not call for Shadow DOM support, so knowledge on the team is currently thin. It will take time for people to ramp up on skills. Not only that, but it isn't part of our weekly testing and we have no plans for it to be at this time, which means we must rely on unit tests as much as possible.
+
+In addition, contributions may occur for a little while until those teams have their needs met. However, it should be assumed that there will be further work to complete the goals as outlined here.
+
+Another concern is that the current approach is, for lack of a better word, hacky. This is because we are accessing and manipulating the Shadow DOM in ways that it wasn't really intended. If we were to rewrite our library today, there are other ways we'd solve these issues which would respect more of the concept of the Shadow DOM and its purpose. What we do here and now may complicate a future where we have different APIs to support this vision of what support would ideally look like.
```

Comment:
I think one way to think about this is supporting different root elements instead of shadow DOM. For example components like modal and popover always treat document.body as their root element. Instead react-aria should provide options to treat any other element as a container element.

Reply:
Possibly, what happens in the case of nested shadow doms though? Would we expect people to `addWindowFocusTracking` equivalent for each root and each hook/function that needs it? Would components need to accept a root element to pass down to hooks? context?

This wouldn't work for closed roots. This is why we want to re-think some of the ways parts of our library work. For example, FocusScope moves focus around for you when it's `contain=true`. This can't move into closed shadow roots, which is why the `video` tag breaks our FocusScope at the moment in certain cases. Instead, if we moved to a sentinel-based approach, we could let the browser handle the tab movement.